scheduler/compileserver.cpp

```diff
@@ -401,3 +410,158 @@ bool CompileServer::blacklisted(const Job *job, const pair<string, string> &envi
     Environments blacklist = job->submitter()->getEnvsForBlacklistedCS(this);
     return find(blacklist.begin(), blacklist.end(), environment) != blacklist.end();
 }
+
+int CompileServer::getInFd() const
+{
+    return m_inFd;
+}
+
+void CompileServer::startInConnectionTest()
+{
+    if (m_noRemote || getConnectionInProgress() || (m_nextConnTime > time(0)))
+    {
+        return;
+    }
+
+    m_inFd = socket(PF_INET, SOCK_STREAM, 0);
+    fcntl(m_inFd, F_SETFL, O_NONBLOCK);
+
+    struct hostent *host = gethostbyname(name.c_str());
+
+    struct sockaddr_in remote_addr;
+    remote_addr.sin_family = AF_INET;
+    remote_addr.sin_port = htons(remotePort());
+    memcpy(&remote_addr.sin_addr.s_addr, host->h_addr_list[0], host->h_length);
+    memset(remote_addr.sin_zero, '\0', sizeof(remote_addr.sin_zero));
+
+    int status = connect(m_inFd, (struct sockaddr *)&remote_addr, sizeof(remote_addr));
+    if(status == 0)
+    {
+        updateInConnectivity(isConnected());
+    }
+    else if (!(errno == EINPROGRESS || errno == EAGAIN))
+    {
+        updateInConnectivity(false);
+    }
+    m_lastConnStartTime=time(0);
+}
+
+void CompileServer::updateInConnectivity(bool acceptingIn)
+{
+    static const time_t time_offset_table[] = {
+        2,    4,    8,    16,    32,
+        64,  128,  256,   512,  1024,
+        2048, 4096
+    };
+    static const size_t table_size = sizeof(time_offset_table);
+
+    //On a successful connection, we should still check back every 1min
+    static const time_t check_back_time = 60;
+
+    if(acceptingIn)
+    {
+        if(!m_acceptingInConnection)
+        {
+            m_acceptingInConnection = true;
+            m_inConnAttempt = 0;
+            trace() << "Client (" << m_nodeName <<
+                " " << name <<
+                ":" << m_remotePort <<
+                ") is accepting incoming connections." << endl;
+        }
+        m_nextConnTime = time(0) + check_back_time;
+        close(m_inFd);
+        m_inFd = -1;
+    }
+    else
+    {
+        if(m_acceptingInConnection)
+        {
+            m_acceptingInConnection = false;
+            trace() << "Client (" << m_nodeName <<
+                " " << name <<
+                ":" << m_remotePort <<
+                ") connected but is not able to accept incoming connections." << endl;
+        }
+        m_nextConnTime = time(0) + time_offset_table[m_inConnAttempt];
+        if(m_inConnAttempt < table_size) m_inConnAttempt++;
+        trace()  << nodeName() << " failed to accept an incoming connection on "
+            << name << ":" << m_remotePort << " attempting again in "
+            << m_nextConnTime - time(0) << " seconds" << endl;
+        close(m_inFd);
+        m_inFd = -1;
+    }
+
+}
+
+void CompileServer::inConnectionResponse(int selectRet, fd_set read_set, fd_set write_set)
```

Comment:
We shouldn't be passing a fd_set to this function.  That select is used is an implementation detail of scheduler.cpp.  Just the FD should be passed in.  We might want to switch to poll from select in the future.